widget/text.go

```diff
@@ -168,13 +168,21 @@ func (t *Text) draw(screen *ebiten.Image) {
 	case TextPositionCenter:
 		p = p.Add(image.Point{0, int((float64(r.Dy()) - t.measurements.boundingBoxHeight) / 2)})
 	case TextPositionEnd:
-		p = p.Add(image.Point{0, int((float64(r.Dy()) - t.measurements.boundingBoxHeight))})
+		p = p.Add(image.Point{0, int(float64(r.Dy()) - t.measurements.boundingBoxHeight)})
 	}
 
 	t.colorList = &datastructures.Stack[color.Color]{}
 	t.colorList.Push(&t.Color)
 
 	for i, line := range t.measurements.lines {
+		ly := int(math.Round(float64(p.Y) + t.measurements.lineHeight*float64(i) + t.measurements.ascent))
+		if ly > screen.Bounds().Max.Y+int(math.Round(t.measurements.ascent)) {
+			return
+		}
+		if ly < - int(math.Round(t.measurements.lineHeight-t.measurements.ascent)) {
```

Comment:
If you change the statement to 
```if ly <= int(math.Round(t.measurements.lineHeight-t.measurements.ascent)) {```
It becomes a bit more aggressive with avoiding drawing too much.

Reply:
So `t.measurements.lineHeight-t.measurements.ascent` is the descent
`ly` is the dot position to consider, meaning the text can extend to the dot position + the descent, which means it does not need to be drawn if  `ly + descent < 0`, or `ly < - descent`.

With the change you’re suggesting, changing the example as follows as well:
```
diff --git a/_examples/widget_demos/textarea/main.go b/_examples/widget_demos/textarea/main.go
index dcc165b..1ba5507 100644
--- a/_examples/widget_demos/textarea/main.go
+++ b/_examples/widget_demos/textarea/main.go
@@ -28,7 +28,6 @@ func main() {

                // the container will use an anchor layout to layout its single child widget
                widget.ContainerOpts.Layout(widget.NewRowLayout(
-                       widget.RowLayoutOpts.Padding(widget.NewInsetsSimple(30)),
                        widget.RowLayoutOpts.Direction(widget.DirectionVertical),
                )),
        )
@@ -91,7 +90,7 @@ func main() {
        //Add text to the end of the textarea
        textarea.AppendText("\nLast Row")
        //Add text to the beginning of the textarea
-       textarea.PrependText("First Row\n")
+       textarea.PrependText("pgq\n")
        //Replace the current text with the new value
        //textarea.SetText("New Value!")
        //Retrieve the current value of the text area text
diff --git a/widget/text.go b/widget/text.go
index 7c43a2c..435f988 100644
--- a/widget/text.go
+++ b/widget/text.go
@@ -179,7 +179,7 @@ func (t *Text) draw(screen *ebiten.Image) {
                if ly > screen.Bounds().Max.Y+int(math.Round(t.measurements.ascent)) {
                        return
                }
-               if ly < - int(math.Round(t.measurements.lineHeight-t.measurements.ascent)) {
+               if ly <= int(math.Round(t.measurements.lineHeight-t.measurements.ascent)) {
                        continue
                }
```
using letters with a bigger descent on the first line, we can see that when scrolling slowly down we go from

![image](https://github.com/ebitenui/ebitenui/assets/2386884/2c4a2a06-6d4c-4948-9618-6b1651b43513)

to

![image](https://github.com/ebitenui/ebitenui/assets/2386884/cbaa62ab-97a8-488d-a071-8600a2f84ef3)

with the bottom of first line disappearing between the two.

Reply:
Ok, I dont know why my comment disappeared but here is my suggestion. We add the following code below your checks to ensure the text isnt rendered if its outside the bounds of the screen, but we add the following checks after it:
```
		if t.widget.parent != nil {
			if ly < t.widget.parent.Rect.Min.Y {
				continue
			}
			if ly-int(math.Round(t.measurements.lineHeight)) > t.widget.parent.Rect.Max.Y {
				return
			}
		}
```
With that check it will check the parent of this text widget and ensure it does not draw outside the bounds of the widget. To make this work properly with the TextArea we have to update textarea.go at line 261 with the following line:

```
	l.text.widget.parent = l.container.GetWidget()
```

That way the text object inside textarea will know the actual bounds that it must be rendered within..